### PR TITLE
[SYCL] Cache and reuse events in the Level Zero plugin.

### PR DESCRIPTION
**Pull request by againull**
Avoid creating new pi_event objects. Reset pi_event objects
and put them in the cache instead of removing.

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -5812,7 +5862,17 @@ pi_result piEventRelease(pi_event Event) {
     Event->CommandData = nullptr;
   }
   if (Event->OwnZeEvent) {
-    ZE_CALL(zeEventDestroy, (Event->ZeEvent));
+    if (!DisableEventsCaching) {
+      ZE_CALL(zeEventHostReset, (Event->ZeEvent));
+      Event->Context->addZeEventToCache(Event->ZeEvent, Event->ZeEventPool,
+                                        Event->isHostVisible(),
+                                        Event->isProfilingEnabled());
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5869]: Can we maybe cache the PI event instead of re-creating from L0 components?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -5812,7 +5862,17 @@ pi_result piEventRelease(pi_event Event) {
...
+    } else {
+      ZE_CALL(zeEventDestroy, (Event->ZeEvent));
+      auto Context = Event->Context;
+      if (auto Res = Context->decrementUnreleasedEventsInPool(Event))
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5873]: The decrementUnreleasedEventsInPool is not being called when caching. Does it not lead to any problems with pools, like leaks, for example? Also, it looks that now we don't really need to cache pools since they never get empty, and thus pools not returned to the pool cache, right?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Good call, thanks, fixed.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5897]: Good catch. 

Indeed when caching is disabled we remove full pools from cache and then add them back if event gets released and available slot appears in the pool. So eventually all pools end up being in the cache (when all events are destroyed and decrementUnreleasedEventsInPool is called for each of them) and we destroy all pool in context destruction.

When event caching is enabled we still need to put pools in the cache because we need to keep creating events in the same pool while there are available slots and keep track of all created pools to destroy them during context destruction. But we don't need to remove full pool from the cache, we just need to keep all full pools at the tail of the cache, pool with available slots is always the first pool in the cache. And just destroy all pools from the cache during Context destruction and there are no leaks.
Correct, thus we don't call decrementUnreleasedEventsInPool when event caching is enabled because we don't destroy events in scope of context and we don't need to add pools back to the cache.

**Comment by againull** [comment]
CUDA backend failures are unrelated.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5897]: Added test for this in https://github.com/intel/llvm-test-suite/pull/1121

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 3]
```diff
@@ -276,6 +276,8 @@ template <class T> struct ZeCache : private T {
 struct ReferenceCounter {
   ReferenceCounter(pi_uint32 InitVal) : RefCount{InitVal} {}
 
+  void reset(pi_uint32 InitVal) { RefCount = InitVal; }
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:279]: this is strange that it takes an arbitrary integer (I think you only call it with "1")
why is this needed?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5897]: Sounds good.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 3 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: I agree, fixed.

**Review by smaslov-intel — APPROVED** [review]
